### DISCUSSION TITLE
Identity and Events

### DISCUSSION THREAD

**Discussion by afflom** [emporous/community#4] [Design]
## EDIT to this is below in the comments

Hi all! 

@jpower432 and I were talking about how we are designing identity and "events" in UOR. She asked me to write a bit here to assist with onboarding interested people into these concepts. 

Let's start with the anatomy of a UOR collection:

 A Collection is a grouping of content which is arranged into a DAG. One of the nodes in a Collection is a link to the root node of a separate and special Collection which has attribute declarations and information. This special Collection is referred to as Schema. Within the Schema, there is a node which is a link to the root node of another separate and special Collection. This special Collection that is linked to from the Schema is called the Event Engine. Event Engines are the application logic that gets associated with content in UOR. 

Here is a drawn representation of these components:
![image](https://user-images.githubusercontent.com/30676559/186984249-2f68581b-92ff-4730-acbe-f386d1dd3ac4.png)

An Event Engine could be a web server, an AI model registry, a Terraform binary... It can be anything that provides a meaningful interaction with the content of the root Collection. I'm not going to go into the actual process flow of UOR or the nitty gritty of how the Event Engine interacts with the root collection and the calling Schema here, but for the purposes of this post, it is sufficient to know that an Event Engine is the application logic that provides a meaningful interaction with the contents of a Collection. 

Here is where we get to Identity. A great point of reference here is Kubernetes. Kube context is the caller's identity in that system. 

In UOR, we've been known to say "everything is an object", which actually means, "everything is a Collection". In UOR, a user is represented by a User Agent. That User Agent is actually the Event Engine of the Identity Collection. Within the Identity Collection, exists resources like the users' preferences (dot files) and secrets. The Schema of the Identity Collection is the Identity Schema. 

When a user uses the UOR CLI to interact with a referenced collection, the user is providing instructions for their User Agent to perform interactions with the referenced Collection on the user's behalf. If needed, the User Agent references content from the user's Identity Collection, such as preferences and secrets. 

Ok, so I'll stop there. Hit me with your questions or comments.

Related: #5 and #6

**Comment by jpower432** [comment]
This issue on the client will come into play, I suspect, for the Identity pieces. https://github.com/uor-framework/uor-client-go/issues/69. We would have to support attribute encryption and encryption on the artifact contents.

**Comment by afflom** [comment]
# Update to this:
@jpower432 and I were talking about this and it is not ideal to import the Event Engine through the schema as the diagram in the OP depicts. We have a solution for this that I will link this comment to its discussion when I post it.
